File: LeetCode/297. Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        result = []
        def dfs(node):
            if not node:
                result.append("N")
                return
            result.append(str(node.val))
            dfs(node.left)
            dfs(node.right)
        dfs(root)
        
        return ",".join(result)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        self.i = 0
        vals = data.split(",")
        def dfs():
            if vals[self.i] == 'N':
                self.i+=1
                return
            n = TreeNode(int(vals[self.i]))
            self.i+=1
            n.left = dfs()
            n.right = dfs()
            return n
            
        return dfs()

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: LeetCode/297. Serialize and Deserialize Binary Tree.go
// https://leetcode.com/problems/serialize-and-deserialize-binary-tree/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

import (
    "strings"
)

var (
    resList []string
    counter int
)

type Codec struct {
    
}

func Constructor() Codec {
    return Codec{} 
}

// Serializes a tree to a single string.
func (this *Codec) serialize(root *TreeNode) string {
    resList = make([]string, 0)
    
    serializerDFS(root)
    return strings.Join(resList, ",")
}

func serializerDFS(root *TreeNode) {
    if root == nil {
        resList = append(resList, "N")
        return
    }
     
    resList = append(resList, strconv.FormatInt(int64(root.Val), 10))
    serializerDFS(root.Left)
    serializerDFS(root.Right)
    return
}

// Deserializes your encoded data to tree.
func (this *Codec) deserialize(data string) *TreeNode {
    s := strings.Split(data, ",") 
    counter = 0
    result := deserDFS(s)
    return result
}

func deserDFS(vals []string) *TreeNode {
    if vals[counter] == "N" {
        counter+=1
        return nil
    }
    var result TreeNode
    if i, err := strconv.ParseInt(vals[counter], 10, 64); err == nil {
        result.Val = int(i)
    }
    counter+=1
    result.Left = deserDFS(vals)
    result.Right = deserDFS(vals)
    return &result
}


/**
 * Your Codec object will be instantiated and called as such:
 * ser := Constructor();
 * deser := Constructor();
 * data := ser.serialize(root);
 * ans := deser.deserialize(data);
 */
